src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,176 @@
+            #
+            # In the future if we know which callback failed, we can
+            # commit only up to that point and retry on the remaining
+            # messages.
+            raise Exception(f"didn't get all the callbacks: {messages_left} left")
+
+        self.__messages.append(message)
+
+    def _callback(self, error: Any, message: Any) -> None:
+        if error is not None:
+            raise Exception(error.str())
+
+    def close(self) -> None:
+        self.__closed = True
+
+        if not self.__messages:
+            return
+
+        # if we have successfully produced messages to the snuba-metrics topic
+        # then enque task to send a slimmed down payload to the product metrics data model.
+        # TODO(meredith): once we know more about what the product data model needs
+        # adjust payload to send the necessary data
+        # messages = [
+        #     {"tags": m.payload["tags"], "name": m.payload["name"], "org_id": m.payload["org_id"]}
+        #     for m in self.__messages
+        # ]
+        # process_indexed_metrics.apply_async(kwargs={"messages": messages})
+        self.__messages = []
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        pass
+
+
+def flush_batch():
+    return FlushBatchStep()
+
+
+@functools.lru_cache(maxsize=10)
+def get_indexer():
+    from sentry.sentry_metrics import indexer
+
+    return indexer
+
+
+def process_messages(messages: Any) -> Sequence[MutableMapping[str, Any]]:
+    indexer = get_indexer()
+
+    strings = set()
+    parsed_messages = [
+        json.loads(message.payload.value, use_rapid_json=True) for message in messages
+    ]
+
+    for message in parsed_messages:
+        metric_name = message["name"]
+        tags = message.get("tags", {})
+
+        parsed_strings = {
+            metric_name,
+            *tags.keys(),
+            *tags.values(),
+        }
+        strings.update(parsed_strings)
+
+    mapping = indexer.bulk_record(list(strings))  # type: ignore
+
+    for i, message in enumerate(parsed_messages):
+        metric_name = parsed_messages[i]["name"]
+        tags = parsed_messages[i].get("tags", {})
+
+        new_tags = {mapping[k]: mapping[v] for k, v in tags.items()}
+
+        message["tags"] = new_tags
+        message["metric_id"] = mapping[metric_name]
+        message["retention_days"] = 90
+
+    return parsed_messages
+
+
+def get_streaming_metrics_consumer(topic: str, **options: Dict[str, str]) -> StreamProcessor:
+    DEFAULT_BLOCK_SIZE = int(32 * 1e6)
+    processing_factory = KafkaConsumerStrategyFactory(
+        None,
+        process_messages,
+        flush_batch,
+        max_batch_size=4,
+        max_batch_time=1000.0,
+        processes=2,
+        input_block_size=DEFAULT_BLOCK_SIZE,
+        output_block_size=DEFAULT_BLOCK_SIZE,
+        initialize_parallel_transform=initializer,
+        batch_execute=True,
+    )
```

Comment:
You should not need a KafkaConsumerStrategyFactory for your use case. 
The KafkaConsumerStrategyFactory is meant for a scenario where you accumulate messages in batches and they are then flushed in batches.
This is good when writing on clickhouse, but your use case does not need it.
Your use case does this:
- poll message
- batch messages
- process batch
- un-batch messages
- produce message
- schedule celery task.

The last two steps are not a batch collector, also, since the kafka producer works asynchronously, having a collector step that works in batches is counter productive as you will be waiting when you flush the messages.

Let's simplify the problem and pretend that what you had to do was this:
- poll message
- transform message
- produce message on kafka

In this case you could just use two simple steps. The first would do the transformation, the second would produce and, critically, would not wait to flush. 
The second step can produce on the `submit` method and store the callback.
Every time the processor calls `poll` you could check how many callbacks have returned positively. and after a given number of them commit on Kafka.
You would never be waiting on a callback or call flush on the producer (which is a waiting operation).

You can use the `MessageRejected` exception the `submit` method can raise to cause back pressure in case you have too many callbacks you are waiting onto.